### PR TITLE
Upgrade `k8s-openapi` to 0.16

### PR DESCRIPTION
**Pull request by clux**
- [x] `cargo upgrade k8s-openapi --workspace`
- [x] remove references to api resources gone in 1.25
- [x] bump our MK8SV and local pin of k8s-openapi to `v1_25` using `just bump-k8s`
- [x] fix k3d CI issue
- [x] reproduce v1_25 issue #997 (it's worse than expected)
- [x] report findings upstream  https://github.com/kubernetes/kubernetes/issues/111985
- [x] ~~await resolution upstream~~ or do an ugly workaround... (see last comment)

TL;DR: upstream bug is bad, breaking, and we need to do something to support v1_25, but we cannot detect it, so it needs to be user-opt in.

### CODE REVIEW

**Comment by clux** [comment]
Integration test failure from an integration test in `derived_resource_queriable_and_has_subresources`:
https://github.com/kube-rs/kube-rs/blob/master/kube/src/lib.rs#L234-L303

with

> Error: Api(ErrorResponse { status: "Failure", message: "Object 'Kind' is missing in '{}'", reason: "", code: 500 })

looks veeery similar to the original bug, but it's actually happening on `get_status`...

**Comment by clux** [comment]
`Error: Api(ErrorResponse { status: "Failure", message: "Object 'Kind' is missing in '{}'", reason: "", code: 500 })`

happens exactly on the 
```rust
foos.delete_collection(&DeleteParams::default(), &Default::default()).await?
```
line. so we can reproduce Arnav's issue on k3d locally with k3d flag: `--image=rancher/k3s:v1.25.0-k3s1-amd64`

**Comment by clux** [comment]
Ok, I think the only thing we NEED to do here is to force in the kind and apiVersion of `DeleteParams`:

```go
apiVersion: meta.k8s.io/v1
kind: DeleteOptions
```
to the output. Never done this before, but how hard can it be.

**Comment by clux** [comment]
Well, great. Serializing with kind/api_version fails in Kubernetes v1.20 (our MK8SV) with:
> Error: Api(ErrorResponse { status: "Failure", message: "no kind \"DeleteOptions\" is registered for version \"meta.k8s.io/v1\" in scheme \"k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:52\"", reason: "", code: 500 })

**Comment by clux** [comment]
Guess it could be wrapped in an [`k8s_openapi::k8s_if_ge_1_25!`](https://docs.rs/k8s-openapi/latest/k8s_openapi/#macros) check... Will look at it more tomorrow.

**Comment by clux** [comment]
Well, this is really quite bad. It's not even trivial for users to deal with this because it effectively forces the user to deal with a **zero version skew** (against our [versioning policies](https://kube.rs/kubernetes-version/#version-skew)).

We have two error cases:

###  Serializing kind/apiVersion breaks on Kubernetes < 1.25 
`api.delete_collection` returns:
```
(Error: Api(ErrorResponse { status: "Failure", message: "no kind \"DeleteOptions\" is registered for version \"meta.k8s.io/v1\" in scheme \"k8s.io/apimachinery@v1.24.1-k3s1/pkg/runtime/scheme.go:100\"", reason: "", code: 500 })
```

### Not serializing kind/apiVersion breaks on Kubernetes >= 1.25
`api.delete_collection` returns:
```
Error: Api(ErrorResponse { status: "Failure", message: "Object 'Kind' is missing in '{}'", reason: "", code: 500 })
```

So the user gotta pick the `v1_25` feature just as they upgrade to it...
we also would need to propagate `k8s-openapi` features in integration tests for just this.

I don't see any good solutions for this at the moment. Comments welcome.

**Comment by clux** [comment]
Have reported findings upstream in https://github.com/kubernetes/kubernetes/issues/111985 .

In the mean time, **options** as I see are:

1. release with the caveat that v1_25 is likely broken for delete_collection calls
2. release with zero-version-skew requirement and do a (currently untestable) `cfg(feature = "v1_25")` in the middle of core
3. hold off releasing until upstream is dealt with

I don't like 3, because historically it is always very confusing for users to have to pick an older k8s-openapi.
I don't like 2, as it's super awkward, gives the impression that all is good, when it really breaks the version skew requirements (which people definitely do not expect to break on them during cluster upgrades).

So Leaning towards 1 with a caveated release and hope pressure can help upstream.

**Comment by clux** [comment]
Ok, have added an **unstable evar** to track this because there's no good way to support v1_25 without the user making a choice here.

Why?
- if we detect Kubernetes version via k8s-openapi features, it's still broken if user is running on a newer cluster
- if we detect Kubernetes version via an api call, it might be broken when/if Kubernetes actually resolves this

Thus, we need to make this an opt in thing. Either via an unstable EVAR or via an unstable feature.
Went with an evar rather than a feature because it's less taxing on our setup, and it doesn't break when/if upstream fixes the bug and we have to remove the feature.

Planning on removing the hacky evar in `api.delete_collection` when upstream fixes the bug.
Also planning on communicating that you do need to set `KUBE_UNSTABLE_V1_25_DELETE` evar when using kube with Kubernetes 1.25 because upstream has seriously fucked up here.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -144,25 +144,30 @@ jobs:
...
+      - name: Workaround upstream Kubernetes 1.25 bug
+        if: matrix.k8s == 'latest'
+        run: echo "KUBE_UNSTABLE_V1_25_DELETE=1" >> $GITHUB_ENV
 
       # Real CI work starts here
       - name: Build workspace
         run: cargo build
 
       # Run the equivalent of `just integration`
-      - name: Run all default features integration library tests
```
Comment (kazk) [inline .github/workflows/ci.yml:154]: Why was this name removed?

File: kube-core/src/request.rs [hunk 2]
```diff
@@ -148,9 +148,33 @@ impl Request {
...
+            }
+            let kind_wrapped_dp = KindWrapper {
+                api_version: "meta.k8s.io/v1",
+                kind: "DeleteOptions",
+                data: serde_json::to_value(&dp).map_err(Error::SerializeBody)?,
+            };
+            serde_json::to_vec(&kind_wrapped_dp).map_err(Error::SerializeBody)?
+        } else {
+            serde_json::to_vec(&dp).map_err(Error::SerializeBody)?
+        };
```
Comment (kazk) [inline kube-core/src/request.rs:174]: This sucks, but I can't think of a way either :(

---

Maybe add a special case for the default `dp` like `k8s_openapi`? 

1. If `dp` is the default, send `vec![]`
2. Otherwise, if the evar is defined, wrap it
3. Otherwise, serialize as is

Comment (clux) [reply kube-core/src/request.rs]: Done in https://github.com/kube-rs/kube-rs/pull/1008/commits/f44041560e350b2276e44ba9632ee3455a3adf31.
This makes `delete_collection` slightly more resilient on 1.25 (but still breaks without the evar when changing delete params).

File: .github/workflows/ci.yml (hunk 1 above)
Comment (clux) [reply .github/workflows/ci.yml]: Accident when refactoring from feature-flag to evar. Fixed now. Thanks!

**Comment by teozkr** [comment]
Honestly I'd just say upstream K8s is broken here and let them fix it (option 1). Especially since it looks like they're already working on a fix, and that I'd wager that there isn't a single K8s cluster in the world that is _only_ running kube-rs clients.

**Comment by clux** [comment]
> Honestly I'd just say upstream K8s is broken here and let them fix it (option 1). Especially since it looks like they're already working on a fix, and that I'd wager that there isn't a single K8s cluster in the world that is _only_ running kube-rs clients.

It's not too different from what we are doing now tbh. We don't have to announce the evar as a workaround, but we do need to work around it to keep CI happy for now.

**Comment by kazk** [comment]
> we should really have the workaround there for us to at least keep CI happy.

Does it still fail with https://github.com/kube-rs/kube-rs/commit/f44041560e350b2276e44ba9632ee3455a3adf31 ?

I think integration tests are all using the default.

**Comment by clux** [comment]
> Does it still fail with [f440415](https://github.com/kube-rs/kube-rs/commit/f44041560e350b2276e44ba9632ee3455a3adf31) ?
> 
> I think integration tests are all using the default.

Think the [e2e test which is using background deletion](https://github.com/kube-rs/kube-rs/blob/master/e2e/job.rs#L45).

**Comment by kazk** [comment]
https://github.com/kube-rs/kube-rs/blob/cc2a2b9a9fa27167f988474d14a925cb8a96afc8/e2e/job.rs?rgh-link-date=2022-09-20T18%3A09%3A54Z#L44-L45

So `delete` is also affected?

Don't we need to work around this?
https://github.com/kube-rs/kube-rs/blob/cc2a2b9a9fa27167f988474d14a925cb8a96afc8/kube-core/src/request.rs#L126-L134

But apparently it's already passing the tests.

**Comment by clux** [comment]
Ah, sorry, I am wrong.

- delete is not affected (just tested)
- e2e does not fail on this in 1.25 (because it's using delete)

**Comment by clux** [comment]
So we can technically remove all this evar gunk for ci :thinking: 

Personally, I feel like having valid work around for an actual release is useful here. Otherwise users have no way around this it if they are testing with v1_25 - save for recreating their cluster with an older version - and we can take out the hack when they fix it upstream without much ceremony. But it is an awkward evar lookup on the other hand.

Will remove the workaround if both of you feel the same.

File: kube-core/src/request.rs [hunk 3]
```diff
@@ -148,9 +148,35 @@ impl Request {
             qp.append_pair("labelSelector", labels);
         }
         let urlstr = qp.finish();
-        let body = serde_json::to_vec(&dp).map_err(Error::SerializeBody)?;
+
+        let data = if dp.is_default() {
+            vec![] // default serialize needs to be empty body
+        } else if std::env::var("KUBE_UNSTABLE_V1_25_DELETE").is_ok() {
+            // Empty delete params triggers an error in kubernetes v1.25
+            // https://github.com/kubernetes/kubernetes/issues/111985
```
[the remark this answers is not in the document]
Comment (clux) [reply kube-core/src/request.rs:156]: how about
```
            // The kube-apiserver in Kubernetes v1.25.0 has broken deserialization of non-default delete options
            // So we have to wrap our DeleteParams inside a kind/apiVersion struct to workaround it
            // https://github.com/kubernetes/kubernetes/issues/111985
            // This workaround does not work on older Kubernetes versions so it has to be opt-in.

            // This is a hack that we plan on removing in the next kube release
            // when upstream hopefully resolves the issue.
```

[the remark this answers is not in the document]
Comment (jesperpedersen) [reply kube-core/src/request.rs:156]: Personally, I would wait.

`kind 0.15.0` only has a 1.25.0 image and 1.25.1 is already out. Yes, it will be a while before a proper fix exists, but putting work-arounds in place would just add additional work.

Time box it (say `1.25.3`), and if not resolved - merge, and release.

[the remark this answers is not in the document]
Comment (kazk) [reply kube-core/src/request.rs:156]: Yeah, that's much better :+1:
The bug is much worse than I thought.

[the remark this answers is not in the document]
Comment (kazk) [reply kube-core/src/request.rs]: I think we want to release it because requiring users to use older `k8s_openapi` is often confusing.

I don't know how many users will be testing against 1.25 anytime soon, so this might not be worth including. This is convenient, but we can document it, and users can always fork and add it themselves to test if they'd like.

[the remark this answers is not in the document]
Comment (jesperpedersen) [reply kube-core/src/request.rs]: Sure :+1:

**Comment by clux** [comment]
Anyone for an approval for this?

**Comment by kazk** [comment]
I'm still debating whether this workaround should be included or not.

Options:

1. Include this workaround and let users test against 1.25 with evar
2. Don't include this workaround, but mention this in release note, so users can create a temporary fork with it to test against 1.25
3. Don't include this workaround, but create a branch that users can use to test against 1.25

1 is most convenient for users, but I'd prefer not to include this hack, especially because most users won't be using 1.25.  
2 may be annoying for users, but I don't think many will need 1.25 now.
3 seems like a nice middle ground.

I'm leaning towards not including (3).

**Comment by clux** [comment]
I really don't think it is a big deal either way, I just don't want us to say we don't support a fully released Kubernetes version because we don't think it's needed yet. That sends a bad signal.

Happy to move the workaround out to a branch though, that is workable for keen users. And since both of you have reservations about including it, it'll at least be less controversial :-)

**Review by kazk — APPROVED** [review]

**Comment by clux** [comment]
Released in 0.75.0.

Release notes are currently stashed in the [draft release for 0.75.0](https://github.com/kube-rs/kube-rs/releases/tag/untagged-6315459bb32156d4db85). Will publish that tomorrow (with updated code links) at some point after the [docs.rs build completes](https://docs.rs/releases/queue). Tried to highlight on user actions needed and made a fourth point for all the improvements.
